Variant write error on Dataframe path

### Bug Description

**What happened:**

**What you expected:**

**Steps to reproduce:**
```scala
test("Test StructType with hudi_type=VARIANT metadata crashes parseTypeDescriptor") {
    // Regression test: a StructType field with hudi_type=VARIANT metadata in the DataFrame API
    // used to crash in HoodieSparkSchemaConverters because the BLOB guard condition called
    // parseTypeDescriptor() which threw for non-custom logical types like VARIANT.
    assume(HoodieSparkUtils.gteqSpark4_0, "Variant type requires Spark 4.0 or higher")

    withTempDir { tmp =>
      val variantMetadata = new MetadataBuilder()
        .putString(HoodieSchema.TYPE_METADATA_FIELD, "VARIANT")
        .build()

      val variantStruct = StructType(Seq(
        StructField("metadata", BinaryType, nullable = false),
        StructField("value", BinaryType, nullable = false)
      ))

      val schema = StructType(Seq(
        StructField("id", LongType, nullable = false),
        StructField("name", StringType),
        StructField("variant_data", variantStruct, nullable = false, metadata = variantMetadata),
        StructField("ts", LongType)
      ))

      val data = Seq(
        Row(1L, "row1", Row(Array[Byte](1, 0), """{"key":"value1"}""".getBytes), 1000L)
      )
      val df = spark.createDataFrame(spark.sparkContext.parallelize(data), schema)

      // This should NOT crash with "parseTypeDescriptor only supports custom logical types, got: VARIANT".
      // The struct+metadata VARIANT is treated as a regular struct (the native VariantType path is separate).
      df.write.format("hudi")
        .option("hoodie.table.name", "variant_struct_test")
        .option("hoodie.datasource.write.recordkey.field", "id")
        .option("hoodie.datasource.write.precombine.field", "ts")
        .mode(SaveMode.Overwrite)
        .save(tmp.getCanonicalPath)

      // Verify the data was written and can be read back
      val readDf = spark.read.format("hudi").load(tmp.getCanonicalPath)
      assert(readDf.count() == 1)
      assert(readDf.select("id", "name").collect()(0).getLong(0) == 1L)
    }
  }
```


### Environment

**Hudi version:**
**Query engine:** (Spark/Flink/Trino etc)
**Relevant configs:**

### Logs and Stack Trace

```
Failed to convert struct type to HoodieSchema: StructType(StructField(id,LongType,false),StructField(name,StringType,true),StructField(variant_data,StructType(StructField(metadata,BinaryType,false),StructField(value,BinaryType,false)),false),StructField(ts,LongType,true))
org.apache.hudi.internal.schema.HoodieSchemaException: Failed to convert struct type to HoodieSchema: StructType(StructField(id,LongType,false),StructField(name,StringType,true),StructField(variant_data,StructType(StructField(metadata,BinaryType,false),StructField(value,BinaryType,false)),false),StructField(ts,LongType,true))
	at org.apache.hudi.HoodieSchemaConversionUtils$.convertStructTypeToHoodieSchema(HoodieSchemaConversionUtils.scala:144)
	at org.apache.hudi.HoodieSchemaConversionUtils$.convertStructTypeToHoodieSchema(HoodieSchemaConversionUtils.scala:121)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.writeInternal(HoodieSparkSqlWriter.scala:371)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$write$1(HoodieSparkSqlWriter.scala:187)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.write(HoodieSparkSqlWriter.scala:205)
	at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:127)
	at org.apache.hudi.DefaultSource.createRelation(DefaultSource.scala:166)
	at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:55)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:79)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:77)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:88)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$eagerlyExecuteCommands$2(QueryExecution.scala:155)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$8(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.withSessionTagsApplied(SQLExecution.scala:272)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$7(SQLExecution.scala:125)
	at org.apache.spark.JobArtifactSet$.withActiveJobArtifactState(JobArtifactSet.scala:94)
	at org.apache.spark.sql.artifact.ArtifactManager.$anonfun$withResources$1(ArtifactManager.scala:112)
	at org.apache.spark.sql.artifact.ArtifactManager.withClassLoaderIfNeeded(ArtifactManager.scala:106)
	at org.apache.spark.sql.artifact.ArtifactManager.withResources(ArtifactManager.scala:111)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$6(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:295)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$1(SQLExecution.scala:124)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:804)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId0(SQLExecution.scala:78)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:237)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$eagerlyExecuteCommands$1(QueryExecution.scala:155)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:654)
	at org.apache.spark.sql.execution.QueryExecution.org$apache$spark$sql$execution$QueryExecution$$eagerlyExecute$1(QueryExecution.scala:154)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$3.applyOrElse(QueryExecution.scala:169)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$3.applyOrElse(QueryExecution.scala:164)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:470)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:86)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:470)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:360)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:356)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:446)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:164)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$lazyCommandExecuted$1(QueryExecution.scala:126)
	at scala.util.Try$.apply(Try.scala:217)
	at org.apache.spark.util.Utils$.doTryWithCallerStacktrace(Utils.scala:1378)
	at org.apache.spark.util.Utils$.getTryWithCallerStacktrace(Utils.scala:1439)
	at org.apache.spark.util.LazyTry.get(LazyTry.scala:58)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:131)
	at org.apache.spark.sql.execution.QueryExecution.assertCommandExecuted(QueryExecution.scala:192)
	at org.apache.spark.sql.classic.DataFrameWriter.runCommand(DataFrameWriter.scala:622)
	at org.apache.spark.sql.classic.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:273)
	at org.apache.spark.sql.classic.DataFrameWriter.saveInternal(DataFrameWriter.scala:241)
	at org.apache.spark.sql.classic.DataFrameWriter.save(DataFrameWriter.scala:118)
	at org.apache.spark.sql.hudi.dml.schema.TestVariantDataType.$anonfun$new$6(TestVariantDataType.scala:133)
	at org.apache.spark.sql.hudi.dml.schema.TestVariantDataType.$anonfun$new$6$adapted(TestVariantDataType.scala:104)
	at org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase.withTempDir(HoodieSparkSqlTestBase.scala:102)
	at org.apache.spark.sql.hudi.dml.schema.TestVariantDataType.$anonfun$new$5(TestVariantDataType.scala:104)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase.$anonfun$test$1(HoodieSparkSqlTestBase.scala:114)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:189)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at org.scalatest.funsuite.AnyFunSuite.withFixture(AnyFunSuite.scala:1562)
	at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:187)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:199)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:199)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:181)
	at org.scalatest.funsuite.AnyFunSuite.runTest(AnyFunSuite.scala:1562)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:232)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:232)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:231)
	at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1562)
	at org.scalatest.Suite.run(Suite.scala:1112)
	at org.scalatest.Suite.run$(Suite.scala:1094)
	at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1562)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:236)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
	at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:236)
	at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:235)
	at org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase.org$scalatest$BeforeAndAfterAll$$super$run(HoodieSparkSqlTestBase.scala:58)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase.run(HoodieSparkSqlTestBase.scala:58)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1320)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1314)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1314)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:993)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:971)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1480)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:971)
	at org.scalatest.tools.Runner$.run(Runner.scala:798)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2or3(ScalaTestRunner.java:43)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:26)
	Suppressed: org.apache.spark.util.Utils$OriginalTryStackTraceException: Full stacktrace of original doTryWithCallerStacktrace caller
		at org.apache.hudi.HoodieSchemaConversionUtils$.convertStructTypeToHoodieSchema(HoodieSchemaConversionUtils.scala:144)
		at org.apache.hudi.HoodieSchemaConversionUtils$.convertStructTypeToHoodieSchema(HoodieSchemaConversionUtils.scala:121)
		at org.apache.hudi.HoodieSparkSqlWriterInternal.writeInternal(HoodieSparkSqlWriter.scala:371)
		at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$write$1(HoodieSparkSqlWriter.scala:187)
		at org.apache.hudi.HoodieSparkSqlWriterInternal.write(HoodieSparkSqlWriter.scala:205)
		at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:127)
		at org.apache.hudi.DefaultSource.createRelation(DefaultSource.scala:166)
		at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:55)
		at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:79)
		at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:77)
		at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:88)
		at org.apache.spark.sql.execution.QueryExecution.$anonfun$eagerlyExecuteCommands$2(QueryExecution.scala:155)
		at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$8(SQLExecution.scala:163)
		at org.apache.spark.sql.execution.SQLExecution$.withSessionTagsApplied(SQLExecution.scala:272)
		at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$7(SQLExecution.scala:125)
		at org.apache.spark.JobArtifactSet$.withActiveJobArtifactState(JobArtifactSet.scala:94)
		at org.apache.spark.sql.artifact.ArtifactManager.$anonfun$withResources$1(ArtifactManager.scala:112)
		at org.apache.spark.sql.artifact.ArtifactManager.withClassLoaderIfNeeded(ArtifactManager.scala:106)
		at org.apache.spark.sql.artifact.ArtifactManager.withResources(ArtifactManager.scala:111)
		at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$6(SQLExecution.scala:125)
		at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:295)
		at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$1(SQLExecution.scala:124)
		at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:804)
		at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId0(SQLExecution.scala:78)
		at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:237)
		at org.apache.spark.sql.execution.QueryExecution.$anonfun$eagerlyExecuteCommands$1(QueryExecution.scala:155)
		at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:654)
		at org.apache.spark.sql.execution.QueryExecution.org$apache$spark$sql$execution$QueryExecution$$eagerlyExecute$1(QueryExecution.scala:154)
		at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$3.applyOrElse(QueryExecution.scala:169)
		at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$3.applyOrElse(QueryExecution.scala:164)
		at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:470)
		at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:86)
		at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:470)
		at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:37)
		at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:360)
		at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:356)
		at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:37)
		at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:37)
		at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:446)
		at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:164)
		at org.apache.spark.sql.execution.QueryExecution.$anonfun$lazyCommandExecuted$1(QueryExecution.scala:126)
		at scala.util.Try$.apply(Try.scala:217)
		at org.apache.spark.util.Utils$.doTryWithCallerStacktrace(Utils.scala:1378)
		at org.apache.spark.util.LazyTry.tryT$lzycompute(LazyTry.scala:46)
		at org.apache.spark.util.LazyTry.tryT(LazyTry.scala:46)
		... 62 more
Caused by: java.lang.IllegalArgumentException: parseTypeDescriptor only supports custom logical types, got: VARIANT
	at org.apache.hudi.common.schema.HoodieSchema.tokenizeTypeDescriptor(HoodieSchema.java:184)
	at org.apache.hudi.common.schema.HoodieSchema.parseTypeDescriptor(HoodieSchema.java:117)
	at org.apache.spark.sql.avro.HoodieSparkSchemaConverters$.toHoodieTypeNested(HoodieSparkSchemaConverters.scala:140)
	at org.apache.spark.sql.avro.HoodieSparkSchemaConverters$.$anonfun$toHoodieTypeNested$2(HoodieSparkSchemaConverters.scala:161)
	at scala.collection.Iterator$$anon$9.next(Iterator.scala:584)
	at scala.collection.immutable.List.prependedAll(List.scala:156)
	at scala.collection.immutable.List$.from(List.scala:685)
	at scala.collection.immutable.List$.from(List.scala:682)
	at scala.collection.SeqFactory$Delegate.from(Factory.scala:306)
	at scala.collection.immutable.Seq$.from(Seq.scala:42)
	at scala.collection.immutable.Seq$.from(Seq.scala:39)
	at scala.collection.IterableOps.map(Iterable.scala:684)
	at scala.collection.IterableOps.map$(Iterable.scala:684)
	at org.apache.spark.sql.types.StructType.map(StructType.scala:105)
	at org.apache.spark.sql.avro.HoodieSparkSchemaConverters$.toHoodieTypeNested(HoodieSparkSchemaConverters.scala:160)
	at org.apache.spark.sql.avro.HoodieSparkSchemaConverters$.toHoodieType(HoodieSparkSchemaConverters.scala:63)
	at org.apache.hudi.HoodieSchemaConversionUtils$.convertStructTypeToHoodieSchema(HoodieSchemaConversionUtils.scala:139)
	at org.apache.hudi.HoodieSchemaConversionUtils$.convertStructTypeToHoodieSchema(HoodieSchemaConversionUtils.scala:121)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.writeInternal(HoodieSparkSqlWriter.scala:371)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$write$1(HoodieSparkSqlWriter.scala:187)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.write(HoodieSparkSqlWriter.scala:205)
	at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:127)
	at org.apache.hudi.DefaultSource.createRelation(DefaultSource.scala:166)
	at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:55)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:79)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:77)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:88)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$eagerlyExecuteCommands$2(QueryExecution.scala:155)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$8(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.withSessionTagsApplied(SQLExecution.scala:272)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$7(SQLExecution.scala:125)
	at org.apache.spark.JobArtifactSet$.withActiveJobArtifactState(JobArtifactSet.scala:94)
	at org.apache.spark.sql.artifact.ArtifactManager.$anonfun$withResources$1(ArtifactManager.scala:112)
	at org.apache.spark.sql.artifact.ArtifactManager.withClassLoaderIfNeeded(ArtifactManager.scala:106)
	at org.apache.spark.sql.artifact.ArtifactManager.withResources(ArtifactManager.scala:111)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$6(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:295)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId0$1(SQLExecution.scala:124)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:804)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId0(SQLExecution.scala:78)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:237)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$eagerlyExecuteCommands$1(QueryExecution.scala:155)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:654)
	at org.apache.spark.sql.execution.QueryExecution.org$apache$spark$sql$execution$QueryExecution$$eagerlyExecute$1(QueryExecution.scala:154)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$3.applyOrElse(QueryExecution.scala:169)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$3.applyOrElse(QueryExecution.scala:164)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:470)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:86)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:470)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:360)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:356)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:446)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:164)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$lazyCommandExecuted$1(QueryExecution.scala:126)
	at scala.util.Try$.apply(Try.scala:217)
	at org.apache.spark.util.Utils$.doTryWithCallerStacktrace(Utils.scala:1378)
	at org.apache.spark.util.LazyTry.tryT$lzycompute(LazyTry.scala:46)
	at org.apache.spark.util.LazyTry.tryT(LazyTry.scala:46)
	... 62 more
```

--- a/hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/avro/HoodieSparkSchemaConverters.scala
+++ b/hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/avro/HoodieSparkSchemaConverters.scala
@@ -103,7 +103,6 @@ object HoodieSparkSchemaConverters extends SparkAdapterSupport {
       // Complex types
       case ArrayType(elementSparkType, containsNull)
           if metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
-            HoodieSchema.isCustomLogicalTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)) &&
             HoodieSchema.parseTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.VECTOR =>
         if (depth > 1) {
           throw new HoodieSchemaException(
@@ -138,11 +137,16 @@ object HoodieSparkSchemaConverters extends SparkAdapterSupport {
         HoodieSchema.createMap(valueSchema)
 
       case blobStruct: StructType if metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
-        HoodieSchema.isCustomLogicalTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)) &&
         HoodieSchema.parseTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.BLOB =>
         // Validate blob structure before accepting
         validateBlobStructure(blobStruct)
         HoodieSchema.createBlob()
+
+      case variantStruct: StructType if metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
+        HoodieSchema.parseTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.VARIANT =>
+        validateVariantStructure(variantStruct)
+        HoodieSchema.createVariant(recordName, nameSpace, null)
+
       case st: StructType =>
         val childNameSpace = if (nameSpace != "") s"$nameSpace.$recordName" else recordName
 
@@ -354,6 +358,33 @@ object HoodieSparkSchemaConverters extends SparkAdapterSupport {
       throw new IllegalArgumentException(
         s"""Invalid blob schema structure. Expected schema:
            |${expectedBlobStructType.toDDL}
+           |Got schema:
+           |${structType.toDDL}""".stripMargin)
+    }
+  }
+
+  private lazy val expectedVariantStructType: StructType = {
+    val metadataField = StructField(HoodieSchema.Variant.VARIANT_METADATA_FIELD, BinaryType, nullable = false)
+    val valueField = StructField(HoodieSchema.Variant.VARIANT_VALUE_FIELD, BinaryType, nullable = false)
+    StructType(Seq(metadataField, valueField))
+  }
+
+  /**
+   * Validates that a StructType matches the expected unshredded variant schema
+   * (two non-null {@code BinaryType} fields: {@code metadata} and {@code value}).
+   *
+   * @param structType the StructType to validate
+   * @throws IllegalArgumentException if the structure does not match the expected variant schema
+   */
+  private def validateVariantStructure(structType: StructType): Unit = {
+    val fieldsByName = structType.fields.map(f => f.name -> f).toMap
+    val ok = structType.length == 2 &&
+      fieldsByName.get(HoodieSchema.Variant.VARIANT_METADATA_FIELD).exists(f => f.dataType == BinaryType && !f.nullable) &&
+      fieldsByName.get(HoodieSchema.Variant.VARIANT_VALUE_FIELD).exists(f => f.dataType == BinaryType && !f.nullable)
+    if (!ok) {
+      throw new IllegalArgumentException(
+        s"""Invalid variant schema structure. Expected schema:
+           |${expectedVariantStructType.toDDL}
            |Got schema:
            |${structType.toDDL}""".stripMargin)
     }

--- a/hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/avro/HoodieSparkSchemaConverters.scala
+++ b/hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/avro/HoodieSparkSchemaConverters.scala
@@ -103,6 +103,7 @@ object HoodieSparkSchemaConverters extends SparkAdapterSupport {
       // Complex types
       case ArrayType(elementSparkType, containsNull)
           if metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
+            HoodieSchema.isCustomLogicalTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)) &&
             HoodieSchema.parseTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.VECTOR =>
         if (depth > 1) {
           throw new HoodieSchemaException(
@@ -137,6 +138,7 @@ object HoodieSparkSchemaConverters extends SparkAdapterSupport {
         HoodieSchema.createMap(valueSchema)
 
       case blobStruct: StructType if metadata.contains(HoodieSchema.TYPE_METADATA_FIELD) &&
+        HoodieSchema.isCustomLogicalTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)) &&
         HoodieSchema.parseTypeDescriptor(metadata.getString(HoodieSchema.TYPE_METADATA_FIELD)).getType == HoodieSchemaType.BLOB =>
         // Validate blob structure before accepting
         validateBlobStructure(blobStruct)

--- a/hudi-common/src/main/java/org/apache/hudi/common/schema/HoodieSchema.java
+++ b/hudi-common/src/main/java/org/apache/hudi/common/schema/HoodieSchema.java
@@ -190,6 +190,24 @@ public class HoodieSchema implements Serializable {
   private static final Set<HoodieSchemaType> CUSTOM_LOGICAL_TYPES =
       EnumSet.of(HoodieSchemaType.VECTOR, HoodieSchemaType.BLOB);
 
+  /**
+   * Returns true if the given descriptor string represents a custom logical type
+   * that can be parsed by {@link #parseTypeDescriptor(String)}.
+   * This is safe to call in pattern-match guards since it never throws.
+   */
+  public static boolean isCustomLogicalTypeDescriptor(String descriptor) {
+    if (descriptor == null || descriptor.trim().isEmpty()) {
+      return false;
+    }
+    int parenStart = descriptor.indexOf('(');
+    String typeName = (parenStart == -1) ? descriptor.trim() : descriptor.substring(0, parenStart).trim();
+    try {
+      HoodieSchemaType type = HoodieSchemaType.valueOf(typeName.toUpperCase(Locale.ROOT));
+      return CUSTOM_LOGICAL_TYPES.contains(type);
+    } catch (IllegalArgumentException e) {
+      return false;
+    }
+  }
 
   /**
    * Constants for Parquet-style accessor patterns used in nested MAP and ARRAY navigation.

--- a/hudi-common/src/main/java/org/apache/hudi/common/schema/HoodieSchema.java
+++ b/hudi-common/src/main/java/org/apache/hudi/common/schema/HoodieSchema.java
@@ -98,8 +98,8 @@ public class HoodieSchema implements Serializable {
   public static final HoodieSchema NULL_SCHEMA = HoodieSchema.create(HoodieSchemaType.NULL);
   /**
    * Constant to use when attaching type metadata to external schema systems like Spark's StructType.
-   * Stores a parameterized type string for custom Hudi logical types such as VECTOR and BLOB.
-   * Examples: "VECTOR(128)", "VECTOR(512, DOUBLE)", "BLOB".
+   * Stores a parameterized type string for custom Hudi logical types such as VECTOR, BLOB, and VARIANT.
+   * Examples: "VECTOR(128)", "VECTOR(512, DOUBLE)", "BLOB", "VARIANT".
    */
   public static final String TYPE_METADATA_FIELD = "hudi_type";
 
@@ -109,8 +109,8 @@ public class HoodieSchema implements Serializable {
   public static final String VARIANT_TYPE_NAME = VariantLogicalType.VARIANT_LOGICAL_TYPE_NAME;
 
   /**
-   * Parses a type descriptor string for custom Hudi logical types such as VECTOR and BLOB.
-   * Examples: "VECTOR(128)", "VECTOR(512, DOUBLE)", "BLOB".
+   * Parses a type descriptor string for custom Hudi logical types such as VECTOR, BLOB, and VARIANT.
+   * Examples: "VECTOR(128)", "VECTOR(512, DOUBLE)", "BLOB", "VARIANT".
    * Throws for non-custom logical type names.
    */
   public static HoodieSchema parseTypeDescriptor(String descriptor) {
@@ -145,6 +145,12 @@ public class HoodieSchema implements Serializable {
               "BLOB type descriptor does not support parameters, got: " + params);
         }
         return createBlob();
+      case VARIANT:
+        if (!params.isEmpty()) {
+          throw new IllegalArgumentException(
+              "VARIANT type descriptor does not support parameters, got: " + params);
+        }
+        return createVariant();
       default:
         throw new IllegalArgumentException(
             "parseTypeDescriptor only supports custom logical types, got: " + type);
@@ -188,26 +194,7 @@ public class HoodieSchema implements Serializable {
   }
 
   private static final Set<HoodieSchemaType> CUSTOM_LOGICAL_TYPES =
-      EnumSet.of(HoodieSchemaType.VECTOR, HoodieSchemaType.BLOB);
-
-  /**
-   * Returns true if the given descriptor string represents a custom logical type
-   * that can be parsed by {@link #parseTypeDescriptor(String)}.
-   * This is safe to call in pattern-match guards since it never throws.
-   */
-  public static boolean isCustomLogicalTypeDescriptor(String descriptor) {
-    if (descriptor == null || descriptor.trim().isEmpty()) {
-      return false;
-    }
-    int parenStart = descriptor.indexOf('(');
-    String typeName = (parenStart == -1) ? descriptor.trim() : descriptor.substring(0, parenStart).trim();
-    try {
-      HoodieSchemaType type = HoodieSchemaType.valueOf(typeName.toUpperCase(Locale.ROOT));
-      return CUSTOM_LOGICAL_TYPES.contains(type);
-    } catch (IllegalArgumentException e) {
-      return false;
-    }
-  }
+      EnumSet.of(HoodieSchemaType.VECTOR, HoodieSchemaType.BLOB, HoodieSchemaType.VARIANT);
 
   /**
    * Constants for Parquet-style accessor patterns used in nested MAP and ARRAY navigation.

--- a/hudi-common/src/test/java/org/apache/hudi/common/schema/TestHoodieSchema.java
+++ b/hudi-common/src/test/java/org/apache/hudi/common/schema/TestHoodieSchema.java
@@ -2855,6 +2855,25 @@ public class TestHoodieSchema {
   }
 
   @Test
+  public void testParseTypeDescriptorVariant() {
+    HoodieSchema parsed = HoodieSchema.parseTypeDescriptor("VARIANT");
+    assertEquals(HoodieSchemaType.VARIANT, parsed.getType());
+    assertInstanceOf(HoodieSchema.Variant.class, parsed);
+  }
+
+  @Test
+  public void testParseTypeDescriptorVariantCaseInsensitive() {
+    HoodieSchema parsed = HoodieSchema.parseTypeDescriptor("variant");
+    assertEquals(HoodieSchemaType.VARIANT, parsed.getType());
+  }
+
+  @Test
+  public void testParseTypeDescriptorVariantRejectsParameters() {
+    assertThrows(IllegalArgumentException.class, () -> HoodieSchema.parseTypeDescriptor("VARIANT(foo)"));
+    assertThrows(IllegalArgumentException.class, () -> HoodieSchema.parseTypeDescriptor("VARIANT(1, 2)"));
+  }
+
+  @Test
   public void testCreateArrayWithNullableVectorThrows() {
     HoodieSchema vectorSchema = HoodieSchema.createNullable(HoodieSchema.createVector(128));
     HoodieSchemaException ex = assertThrows(HoodieSchemaException.class,

--- a/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/avro/TestSchemaConverters.scala
+++ b/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/avro/TestSchemaConverters.scala
@@ -250,6 +250,51 @@ class TestSchemaConverters extends SparkAdapterSupport {
   }
 
   @Test
+  def testInvalidVariantSchemaWrongFieldCount(): Unit = {
+    val invalidStruct = new StructType(Array[StructField](
+      StructField(HoodieSchema.Variant.VARIANT_METADATA_FIELD, DataTypes.BinaryType, nullable = false)
+    ))
+    assertInvalidVariantSchema(invalidStruct)
+  }
+
+  @Test
+  def testInvalidVariantSchemaNullableMetadataField(): Unit = {
+    val invalidStruct = new StructType(Array[StructField](
+      StructField(HoodieSchema.Variant.VARIANT_METADATA_FIELD, DataTypes.BinaryType, nullable = true),
+      StructField(HoodieSchema.Variant.VARIANT_VALUE_FIELD, DataTypes.BinaryType, nullable = false)
+    ))
+    assertInvalidVariantSchema(invalidStruct)
+  }
+
+  @Test
+  def testInvalidVariantSchemaWrongValueFieldType(): Unit = {
+    val invalidStruct = new StructType(Array[StructField](
+      StructField(HoodieSchema.Variant.VARIANT_METADATA_FIELD, DataTypes.BinaryType, nullable = false),
+      StructField(HoodieSchema.Variant.VARIANT_VALUE_FIELD, DataTypes.StringType, nullable = false)
+    ))
+    assertInvalidVariantSchema(invalidStruct)
+  }
+
+  @Test
+  def testInvalidVariantSchemaWrongFieldNames(): Unit = {
+    val invalidStruct = new StructType(Array[StructField](
+      StructField("foo", DataTypes.BinaryType, nullable = false),
+      StructField("bar", DataTypes.BinaryType, nullable = false)
+    ))
+    assertInvalidVariantSchema(invalidStruct)
+  }
+
+  private def assertInvalidVariantSchema(invalidStruct: StructType): Unit = {
+    val metadata = new MetadataBuilder()
+      .putString(HoodieSchema.TYPE_METADATA_FIELD, HoodieSchemaType.VARIANT.name())
+      .build()
+    val exception = assertThrows(classOf[IllegalArgumentException], () => {
+      HoodieSparkSchemaConverters.toHoodieType(invalidStruct, metadata = metadata)
+    })
+    assertTrue(exception.getMessage.startsWith("Invalid variant schema structure"))
+  }
+
+  @Test
   def testTopLevelVectorStillAllowed(): Unit = {
     val vectorMetadata = new MetadataBuilder()
       .putString(HoodieSchema.TYPE_METADATA_FIELD, "VECTOR(4)")

--- a/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/schema/TestVariantDataType.scala
+++ b/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/schema/TestVariantDataType.scala
@@ -94,10 +94,9 @@ class TestVariantDataType extends HoodieSparkSqlTestBase {
     }
   }
 
-  test("Test StructType with hudi_type=VARIANT metadata does not crash parseTypeDescriptor") {
-    // Regression test: a StructType field with hudi_type=VARIANT metadata in the DataFrame API
-    // used to crash in HoodieSparkSchemaConverters because the BLOB guard condition called
-    // parseTypeDescriptor() which threw for non-custom logical types like VARIANT.
+  test("Test StructType with hudi_type=VARIANT metadata is promoted to VARIANT logical type") {
+    // A StructType field in the DataFrame API tagged with hudi_type=VARIANT is treated as a first-class
+    // VARIANT (like BLOB/VECTOR), not a plain struct. On Spark 4.0+ the column round-trips as native VariantType.
     assume(HoodieSparkUtils.gteqSpark4_0, "Variant type requires Spark 4.0 or higher")
 
     withTempDir { tmp =>
@@ -122,8 +121,6 @@ class TestVariantDataType extends HoodieSparkSqlTestBase {
       )
       val df = spark.createDataFrame(spark.sparkContext.parallelize(data), schema)
 
-      // This should NOT crash with "parseTypeDescriptor only supports custom logical types, got: VARIANT".
-      // The struct+metadata VARIANT is treated as a regular struct (the native VariantType path is separate).
       df.write.format("hudi")
         .option("hoodie.table.name", "variant_struct_test")
         .option("hoodie.datasource.write.recordkey.field", "id")
@@ -131,10 +128,49 @@ class TestVariantDataType extends HoodieSparkSqlTestBase {
         .mode(SaveMode.Overwrite)
         .save(tmp.getCanonicalPath)
 
-      // Verify the data was written and can be read back
       val readDf = spark.read.format("hudi").load(tmp.getCanonicalPath)
+      val readFieldType = readDf.schema("variant_data").dataType
+      assert(readFieldType.typeName == "variant",
+        s"variant_data should round-trip as native VariantType on Spark 4.0+, got $readFieldType")
       assert(readDf.count() == 1)
-      assert(readDf.select("id", "name").collect()(0).getLong(0) == 1L)
+    }
+  }
+
+  test("Test StructType with hudi_type=VARIANT metadata rejects malformed struct") {
+    assume(HoodieSparkUtils.gteqSpark4_0, "Variant type requires Spark 4.0 or higher")
+
+    withTempDir { tmp =>
+      val variantMetadata = new MetadataBuilder()
+        .putString(HoodieSchema.TYPE_METADATA_FIELD, "VARIANT")
+        .build()
+
+      // VARIANT structure must be {metadata: binary, value: binary}; a single string field is malformed.
+      val malformedVariantStruct = StructType(Seq(
+        StructField("wrong_field", StringType, nullable = false)
+      ))
+
+      val schema = StructType(Seq(
+        StructField("id", LongType, nullable = false),
+        StructField("variant_data", malformedVariantStruct, nullable = false, metadata = variantMetadata),
+        StructField("ts", LongType)
+      ))
+
+      val data = Seq(Row(1L, Row("oops"), 1000L))
+      val df = spark.createDataFrame(spark.sparkContext.parallelize(data), schema)
+
+      val ex = intercept[Exception] {
+        df.write.format("hudi")
+          .option("hoodie.table.name", "variant_malformed_test")
+          .option("hoodie.datasource.write.recordkey.field", "id")
+          .option("hoodie.datasource.write.precombine.field", "ts")
+          .mode(SaveMode.Overwrite)
+          .save(tmp.getCanonicalPath)
+      }
+      val causes = Iterator.iterate[Throwable](ex)(e => e.getCause).takeWhile(_ != null).toList
+      assert(causes.exists(c => c.isInstanceOf[IllegalArgumentException]
+        && c.getMessage != null
+        && c.getMessage.contains("Invalid variant schema structure")),
+        s"Expected IllegalArgumentException with 'Invalid variant schema structure', got: ${causes.map(_.getMessage)}")
     }
   }
 

--- a/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/schema/TestVariantDataType.scala
+++ b/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/schema/TestVariantDataType.scala
@@ -20,11 +20,14 @@
 package org.apache.spark.sql.hudi.dml.schema
 
 import org.apache.hudi.HoodieSparkUtils
+import org.apache.hudi.common.schema.HoodieSchema
 import org.apache.hudi.common.testutils.HoodieTestUtils
 import org.apache.hudi.common.util.StringUtils
 import org.apache.hudi.internal.schema.HoodieSchemaException
 
+import org.apache.spark.sql.{Row, SaveMode}
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase
+import org.apache.spark.sql.types._
 
 
 class TestVariantDataType extends HoodieSparkSqlTestBase {
@@ -88,6 +91,50 @@ class TestVariantDataType extends HoodieSparkSqlTestBase {
           Seq(1, "row1", "{\"new_field\":123,\"updated\":true}", 1000)
         )
       })
+    }
+  }
+
+  test("Test StructType with hudi_type=VARIANT metadata does not crash parseTypeDescriptor") {
+    // Regression test: a StructType field with hudi_type=VARIANT metadata in the DataFrame API
+    // used to crash in HoodieSparkSchemaConverters because the BLOB guard condition called
+    // parseTypeDescriptor() which threw for non-custom logical types like VARIANT.
+    assume(HoodieSparkUtils.gteqSpark4_0, "Variant type requires Spark 4.0 or higher")
+
+    withTempDir { tmp =>
+      val variantMetadata = new MetadataBuilder()
+        .putString(HoodieSchema.TYPE_METADATA_FIELD, "VARIANT")
+        .build()
+
+      val variantStruct = StructType(Seq(
+        StructField("metadata", BinaryType, nullable = false),
+        StructField("value", BinaryType, nullable = false)
+      ))
+
+      val schema = StructType(Seq(
+        StructField("id", LongType, nullable = false),
+        StructField("name", StringType),
+        StructField("variant_data", variantStruct, nullable = false, metadata = variantMetadata),
+        StructField("ts", LongType)
+      ))
+
+      val data = Seq(
+        Row(1L, "row1", Row(Array[Byte](1, 0), """{"key":"value1"}""".getBytes), 1000L)
+      )
+      val df = spark.createDataFrame(spark.sparkContext.parallelize(data), schema)
+
+      // This should NOT crash with "parseTypeDescriptor only supports custom logical types, got: VARIANT".
+      // The struct+metadata VARIANT is treated as a regular struct (the native VariantType path is separate).
+      df.write.format("hudi")
+        .option("hoodie.table.name", "variant_struct_test")
+        .option("hoodie.datasource.write.recordkey.field", "id")
+        .option("hoodie.datasource.write.precombine.field", "ts")
+        .mode(SaveMode.Overwrite)
+        .save(tmp.getCanonicalPath)
+
+      // Verify the data was written and can be read back
+      val readDf = spark.read.format("hudi").load(tmp.getCanonicalPath)
+      assert(readDf.count() == 1)
+      assert(readDf.select("id", "name").collect()(0).getLong(0) == 1L)
     }
   }
 

